waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -507,10 +535,40 @@
           (<= attempt max-patch-retries)
           (recur (inc attempt) (get-replica-count scheduler service-id)))))))
 
-(defn- kill-service-instance
+(let [num-deletion-locks 100
+      deletion-locks (repeatedly num-deletion-locks (fn [] (Object.)))]
+  (defn- service-id->deletion-lock
+    "Returns the deletion lock striped by a hash of the service ID."
+    [service-id]
+    (nth deletion-locks (mod (hash service-id) num-deletion-locks))))
```

Comment:
`repeatedly` returns a lazy sequence, which means this `nth` operation has to traverse the 100-element lazy list to get the specified item each time. 😱 If we don't delete the locks altogether (see comment below asking what the lock is accomplishing), then we should convert the sequence to a vector to get fast indexing.